File: app/lib/core/algorithm_registry.rb
# frozen_string_literal: true

require_relative "../algorithms/sorting"
require_relative "../algorithms/searching"
require_relative "../algorithms/k_means"
require_relative "../algorithms/simulated_annealing"
require_relative "../algorithms/genetic_algorithm"

module AlgorithmRegistry
  ALGORITHMS = {
    "quick_sort" => Sorting.method(:quick_sort),
    "merge_sort" => Sorting.method(:merge_sort),
    "bubble_sort" => Sorting.method(:bubble_sort),
    "insertion_sort" => Sorting.method(:insertion_sort),
    "heap_sort" => Sorting.method(:heap_sort),
    "radix_sort" => Sorting.method(:radix_sort),
    "linear_search" => Searching.method(:linear_search),
    "binary_search" => Searching.method(:binary_search),
    "ternary_search" => Searching.method(:ternary_search),
    "exponential_search" => Searching.method(:exponential_search),
    "k_means" => KMeans.method(:k_means),
    "simulated_annealing" => SimulatedAnnealing.method(:simulated_annealing),
    "genetic_algorithm" => GeneticAlgorithm.method(:genetic_algorithm)
  }.freeze

  module_function

  def list
    ALGORITHMS.keys
  end

  def get(name)
    ALGORITHMS[name.to_s.downcase]
  end

  def exists?(name)
    ALGORITHMS.key?(name.to_s.downcase)
  end
end

File: app/js/lib/core/algorithm_registry.js
const Sorting = require('../algorithms/sorting');
const Searching = require('../algorithms/searching');

const algorithms = {
  quick_sort: Sorting.quickSort,
  merge_sort: Sorting.mergeSort,
  bubble_sort: Sorting.bubbleSort,
  insertion_sort: Sorting.insertionSort,
  heap_sort: Sorting.heapSort,
  radix_sort: Sorting.radixSort,
  linear_search: Searching.linearSearch,
  binary_search: Searching.binarySearch,
  ternary_search: Searching.ternarySearch,
  exponential_search: Searching.exponentialSearch
};

function getAlgorithm(name) {
  const algorithm = algorithms[name];
  if (!algorithm) {
    throw new Error(`Unknown algorithm: ${name}`);
  }
  return algorithm;
}

function listAlgorithms() {
  return Object.keys(algorithms);
}

module.exports = { getAlgorithm, listAlgorithms };
